src/components/Tutorial/__tests__/Tutorial.ts

```diff
@@ -0,0 +1,317 @@
+import { getByText, screen } from '@testing-library/dom'
+import userEvent from '@testing-library/user-event'
+import { act } from 'react'
+import { importTextActionCreator as importText } from '../../../actions/importText'
+import { newSubthoughtActionCreator as newSubthought } from '../../../actions/newSubthought'
+import { newThoughtActionCreator as newThought } from '../../../actions/newThought'
+import { tutorialNextActionCreator as tutorialNext } from '../../../actions/tutorialNext'
+import { tutorialStepActionCreator as tutorialStep } from '../../../actions/tutorialStep'
+import {
+  TUTORIAL2_STEP_START,
+  TUTORIAL_STEP_AUTOEXPAND,
+  TUTORIAL_STEP_AUTOEXPAND_EXPAND,
+  TUTORIAL_STEP_SECONDTHOUGHT,
+  TUTORIAL_STEP_SUBTHOUGHT,
+  TUTORIAL_STEP_SUCCESS,
+} from '../../../constants'
+import { cleanupTestApp, createTestAppWithTutorial } from '../../../test-helpers/createTestApp'
+import dispatch from '../../../test-helpers/dispatch'
+import { setCursorFirstMatchActionCreator as setCursorFirstMatch } from '../../../test-helpers/setCursorFirstMatch'
+
+beforeEach(createTestAppWithTutorial)
+afterEach(cleanupTestApp)
+
+describe('Tutorial 1', async () => {
+  describe('step start', () => {
+    it('shows the welcome text', () => {
+      expect(screen.getByText('Welcome to your personal thoughtspace.')).toBeInTheDocument()
+    })
+
+    it('shows the "Next" button, that leads to the first step', async () => {
+      const user = userEvent.setup({ delay: null })
+      await user.click(screen.getByText('Next'))
+      await act(vi.runOnlyPendingTimersAsync)
+      expect(
+        screen.getByText('First, let me show you how to create a new thought', { exact: false }),
+      ).toBeInTheDocument()
+    })
+  })
+
+  it('step first thought - how to create a thought ', async () => {
+    expect(screen.getByText('Hit the Enter key to create a new thought.')).toBeInTheDocument()
+
+    const user = userEvent.setup({ delay: null })
+    await act(() => user.keyboard('{Enter}'))
+    expect(screen.getByText('You did it!')).toBeInTheDocument()
+    expect(screen.getByText('Now type something. Anything will do.')).toBeInTheDocument()
+
+    await dispatch(newThought({ value: 'my first thought' }))
+    await act(vi.runOnlyPendingTimersAsync)
+    expect(screen.getByText('Click the Next button when you are ready to continue.')).toBeInTheDocument()
+  })
+
+  describe('step second thought - create another thought', () => {
+    it('prompts to type some text to the created thought, then congratulates on typing', async () => {
+      await dispatch([
+        newThought({ value: 'already created thought' }),
+        tutorialStep({ value: TUTORIAL_STEP_SECONDTHOUGHT }),
```

Comment:
The PR description says that manual clicking is usually prioritized over store-side creation, which I think I would also prefer. However, there are quite a few places here where we dispatch actions to the store rather than interacting with the page. Is there a reason to do that instead of e.g. clicking the "Next" button like we do elsewhere (`await user.click(screen.getByText('Next'))`)?

Same for e.g. `await dispatch(newThought({ value: 'my first thought' }))` rather than typing in the thought.

Reply:
yeah, I kinda lied in the end...
I had these spans of statements to find the correct thought in the space, click on it, trigger a keyboard shortcut to add a thought by `user.type`, but in the end it felt so pointless, as I found myself the core-est functionality of the app, whilst I was tasked with testing a tutorial. And that added some complexity, removing the clarity I tried to achieve, since it's one of the bigger test files in the project
and it felt like dispatching new thoughts was as productive as clicking on things, as the tutorial's logic relies on redux state a lot.

and as per `user.click(screen.getByText('Next'))`, there are two reasaons:
1. a lot of times we don't have the `next` button, the next step is triggered conditionally from store values
2. the test cases became too dependant on each other. E.g. we remove one test case -- and have a cascade of failures, since we didn't increment the step on the removed case. One could argue it's fine, but the ergonomics of being able to do `it.only` or `it.skip` wherever we want to isolate cases, was too tempting.
And I've kinda tested this `Next` thing in `shows the "Next" button, that leads to the first step` case, and its functionality doesn't differ in other cases

Reply:
Thanks for the explanation. While I can see the desire to simplify the tests, I think we lose something essential by not imitating the user's exact actions. Whenever you dispatch an action directly, you open a gap in the test coverage. How do you know that the user can trigger the action? The only way to ensure that the user can complete the tutorial is by automating the exact steps that the user would take.

> 1. a lot of times we don't have the `next` button, the next step is triggered conditionally from store values

This is a good example of my point. Enabling the conditions for the Next button to appear should be under test coverage. This is acutally the area where the tutorial has experienced the most regressions. If a Redux selector changes unintentionally, then the tutorial step is not testing the correct condition any more, preventing the user from advancing. My main reason for #2313 was to put the tutorial advancement under test coverage. Manually advancing defeats this purpose.

> 2\. the test cases became too dependant on each other. E.g. we remove one test case -- and have a cascade of failures, since we didn't increment the step on the removed case. One could argue it's fine, but the ergonomics of being able to do `it.only` or `it.skip` wherever we want to isolate cases, was too tempting.

The root dependency is that the tutorial steps are dependent on each other. If there is a bug in step 3, then the user naturally cannot advance to step 4. If the test for step 3 could fail but the test for step 4 still pass, that indicates a gap in test coverage.

It's possible to jump to the correct tutorial step at the beginning of each test, but then you would have to recreate the start state exactly as it was at the end of the previous step. Otherwise, there would be a gap in the test coverage.

I hope that helps explain the importance of imitating the user's actions in order to ensure that the tutorial can be completed. If a certain set of actions seem overly complex or difficult to abstract, feel free to consult with Trevin to find a clean solution.

Reply:
yeah, now I totally agree with "next" example, thank you

but still, I feel like in some places(like autoexpand) I shouldn't do manual clicks I feel like, since it's already covered

let me see again which parts are lacking attention

Reply:
The expansion behavior itself is covered in other tests, but we would still want to test that the tutorial can detect that the thought has been collapsed or expanded per the instructions to the user.